README.md
# netflow-traffic-accounter

NetFlow IP Traffic Accounting

This package provides a single class which exports aggregate traffic data of
local networks at a regular interval (by default every 5 minutes).

This wouldn't be possible without the the still-under-development
[node-netflowv9](https://github.com/delian/node-netflowv9) package, whose
author deserves the most credit!

NetFlow version 1,5,7 and 9 are supported.

Support for IPv6 (requiring NetFlow version 9) is not yet supported.

This has been tested with the RouterOS 'Traffic Flow' feature, and at the time
of writing [node-netflowv9](https://github.com/delian/node-netflowv9) has been
tested with Cisco IOS XR.

## Installation
    npm install --save netflow-traffic-accounter

## Usage
    const NetflowTrafficAccounter = require('netflow-traffic-accounter');

    const accounter = new NetflowTrafficAccounter({
        listenPort: 3000,
        exportInterval: 300,
        localNets: [
            "10.10.10.0/24",
            "192.168.0.0/24",
        ],
        ignoreNets: [
            "192.168.1.0/24",
        ],
        netflowOptions: {
            // options here are passed to node-netflowv9
        },
    }, function(data, periodFrom, periodTo) {
        // called every exportInterval seconds with traffic data collected
        // since the last export.

        // periodFrom, periodTo contain Unix timestamps of the collection
        // period's start and end.
    });

Traffic data is presented in the following format:

    {
        "10.10.10.1": {
            "in_bytes": 123123,
            "out_bytes": 123123,
            "in_pkts": 1231,
            "out_pkts": 1231,
        },
        "10.10.10.2": {
            ....
        },
        ...
    }

## Options

The NetflowTrafficAccounter class is intialized with an options object as the
first parameter and the data export callback as the second parameter.

The following options are available in the options object:

**listenPort** - Default: 3000. The port to listen for NetFlow data on.

**exportInterval** - Default: 300. How often to export aggregated traffic data
(in seconds). Data is is exported when the following condition is true:

    currentSystemTime % exportInterval == 0

**localNets** - Default: null. An array of subnets considered local, which we
will account incoming and outgoing traffic for. Can be an array of strings or
of `Netmask` objects from the [Netmask](https://github.com/rs/node-netmask)
package.

**ignoreNets** - Default: null. An array of subnets to ignore. We will not
account traffic between localNets and ignoreNets. Can be an array of strings or
of `Netmask` objects from the [Netmask](https://github.com/rs/node-netmask)
package.

**netflowOptions** - Default: {}. Used as the constructor of `node-netflowv9`'s
`Collector`. See [here](https://github.com/delian/node-netflowv9#Options) for
available options. Useful for proxying netflow traffic to another collector with
the `proxy` option. **Note**: Specifying `port` in `netflowOptions` overrides
our own `listenPort`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.1](https://github.com/mlow/netflow-traffic-accounter/compare/v0.4.0...v0.4.1) (2021-02-12)

### Fixed

- Fix: make \_export() async and await the callback (9d2cc2c5afef15ce4ce9a3d60b77667d5a23866e)
- Ensure an in-progress export is awaited in stop (31774fa52aef4f5349a7c850634d05b1aea3d7ce)

## [0.4.0](https://github.com/mlow/netflow-traffic-accounter/compare/v0.3.1...v0.4.0) (2020-03-31)

### Added

- Add ignoreNets option which allows for not counting traffic between localNets
  hosts and ignoreNets hosts.

### Changed

- No longer account packets where ipv4_next_hop is in local_nets

## [0.3.1](https://github.com/mlow/netflow-traffic-accounter/compare/v0.3.0...v0.3.1) (2019-09-10)

- Fixed erroneous incorrect flow sequence warnings.

## [0.3.0](https://github.com/mlow/netflow-traffic-accounter/compare/v0.2.0...v0.3.0) (2019-09-09)

### Added

- Add checking for duplicate/out-of-sequence Netflow packets. Warning messages
  are emitted for either, duplicate packets are discarded.

## [0.2.0](https://github.com/mlow/netflow-traffic-accounter/compare/v0.1.0...v0.2.0) (2019-09-05)

### Added

- Emit 'export' event as alternative to using a callback.
- Allow passing options directly to `node-netflowv9`'s `Collector`. See
  [here](https://github.com/delian/node-netflowv9#Options) for available options.

## [0.1.0](https://github.com/mlow/netflow-traffic-accounter/tree/v0.1.0) (2019-09-04)

### Added

- Basic support for IPv4 traffic accounting


index.js
const Collector = require("node-netflowv9");
const Netmask = require("netmask").Netmask;
const EventEmitter = require("events");

module.exports = class extends (
  EventEmitter
) {
  constructor(options, callback) {
    super();
    this.listenPort = 3000;
    this.exportInterval = 300;
    this.localNets = null;
    this.ignoreNets = null;
    this.callback = callback;
    this.netflowOptions = {};

    this.records = {};
    this.periodFrom = Math.trunc(Date.now() / 1000);
    this.lastExport = 0;
    this.lastSeq = 0;
    this.lastCount = 0;

    this.lastExportPromise = null;

    if (typeof options === "object") {
      if (options.listenPort) this.listenPort = options.listenPort;
      if (options.exportInterval) this.exportInterval = options.exportInterval;
      if (options.localNets) {
        this.localNets = this.asNetmaskArray(options.localNets);
      }
      if (options.ignoreNets) {
        this.ignoreNets = this.asNetmaskArray(options.ignoreNets);
      }
      if (options.netflowOptions) this.netflowOptions = options.netflowOptions;
    }

    this.collector = Collector({
      port: this.listenPort,
      ...this.netflowOptions,
    });
    this.collector.on("template", (record) => {
      this._checkSequence(record);
    });
    this.collector.on("data", (record) => {
      if (!this._checkSequence(record)) {
        return;
      }

      for (const f of record.flows) {
        const v4src = f.ipv4_src_addr;
        const v4dst = f.ipv4_dst_addr;
        const bytes = f.in_bytes;
        const pkts = f.in_pkts;

        if (this.isLocal(v4dst) && !this.isIgnored(v4src)) {
          this.accountTraffic(v4dst, "in", bytes, pkts);
        } else if (this.isLocal(v4src) && !this.isIgnored(v4dst)) {
          this.accountTraffic(v4src, "out", bytes, pkts);
        }
      }
    });

    this.timer = setInterval(() => {
      const now = Math.trunc(Date.now() / 1000);
      if (now % this.exportInterval == 0 && now > this.lastExport) {
        this.lastExport = now;
        this.lastExportPromise = this._export();
      }
    }, 500);
  }

  asNetmaskArray(arr) {
    return arr.map((net) => {
      if (net instanceof Netmask) {
        return net;
      } else if (typeof net === "string" || net instanceof String) {
        return new Netmask(net);
      }
    });
  }

  accountTraffic(addr, direction, bytes, packets) {
    if (!(addr in this.records)) {
      this.records[addr] = {
        in_bytes: 0,
        out_bytes: 0,
        in_pkts: 0,
        out_pkts: 0,
      };
    }
    this.records[addr][direction + "_bytes"] += bytes;
    this.records[addr][direction + "_pkts"] += packets;
  }

  isLocal(addr) {
    return this.localNets.find((net) => net.contains(addr));
  }

  isIgnored(addr) {
    return this.ignoreNets && this.ignoreNets.find((net) => net.contains(addr));
  }

  async _export() {
    const toExport = this.records;
    this.records = {};
    if (this.callback instanceof Function) {
      await this.callback(toExport, this.periodFrom, this.lastExport);
    }
    this.emit("export", toExport, this.periodFrom, this.lastExport);
    this.periodFrom = this.lastExport;
  }

  _checkSequence(record) {
    const version = record.header.version;
    const seq = record.header.sequence;
    const count = record.header.count;

    if (this.lastCount == 0) {
      this.lastSeq = seq;
      this.lastCount = count;
      return true;
    }

    let expectedSeq = -1;
    switch (version) {
      case 5:
        expectedSeq = this.lastSeq + this.lastCount;
        break;
      case 9:
        expectedSeq = this.lastSeq + 1;
        break;
      default:
        return true;
    }

    if (seq == this.lastSeq) {
      console.warn(
        `WARNING: Duplicate flow sequence ${seq} received, discarding.`
      );
      return false;
    }

    if (seq != expectedSeq) {
      console.warn(
        `WARNING: Flow sequence ${seq} is not the expected ${expectedSeq}.`
      );
    }

    this.lastSeq = seq;
    this.lastCount = count;
    return true;
  }

  async stop() {
    clearInterval(this.timer);
    this.collector.server.close();
    if (this.lastExportPromise) {
      // if there was an in-progress export, await it
      await this.lastExportPromise;
    }

    this.lastExport = Math.trunc(Date.now() / 1000);
    // return one final export's promise
    return this._export();
  }
};
